File: code/java/ch11/src/main/java/com/github/zxh0/luago/state/Arithmetic.java
package com.github.zxh0.luago.state;

import com.github.zxh0.luago.api.ArithOp;
import com.github.zxh0.luago.number.LuaMath;

import java.util.function.DoubleBinaryOperator;
import java.util.function.LongBinaryOperator;

class Arithmetic {

    private static final LongBinaryOperator[] integerOps = {
            (a, b) -> a + b,     // LUA_OPADD
            (a, b) -> a - b,     // LUA_OPSUB
            (a, b) -> a * b,     // LUA_OPMUL
            Math::floorMod,      // LUA_OPMOD
            null,                // LUA_OPPOW
            null,                // LUA_OPDIV
            Math::floorDiv,      // LUA_OPIDIV
            (a, b) -> a & b,     // LUA_OPBAND
            (a, b) -> a | b,     // LUA_OPBOR
            (a, b) -> a ^ b,     // LUA_OPBXOR
            LuaMath::shiftLeft,  // LUA_OPSHL
            LuaMath::shiftRight, // LUA_OPSHR
            (a, b) -> -a,        // LUA_OPUNM
            (a, b) -> ~a,        // LUA_OPBNOT
    };

    private static final DoubleBinaryOperator[] floatOps = {
            (a, b) -> a + b,   // LUA_OPADD
            (a, b) -> a - b,   // LUA_OPSUB
            (a, b) -> a * b,   // LUA_OPMUL
            LuaMath::floorMod, // LUA_OPMOD
            Math::pow,         // LUA_OPPOW
            (a, b) -> a / b,   // LUA_OPDIV
            LuaMath::floorDiv, // LUA_OPIDIV
            null,              // LUA_OPBAND
            null,              // LUA_OPBOR
            null,              // LUA_OPBXOR
            null,              // LUA_OPSHL
            null,              // LUA_OPSHR
            (a, b) -> -a,      // LUA_OPUNM
            null,              // LUA_OPBNOT
    };

    static Object arith(Object a, Object b, ArithOp op) {
        LongBinaryOperator integerFunc = integerOps[op.ordinal()];
        DoubleBinaryOperator floatFunc = floatOps[op.ordinal()];

        if (floatFunc == null) { // bitwise
            Long x = LuaValue.toInteger(a);
            if (x != null) {
                Long y = LuaValue.toInteger(b);
                if (y != null) {
                    return integerFunc.applyAsLong(x, y);
                }
            }
        } else { // arith
            if (integerFunc != null) { // add,sub,mul,mod,idiv,unm
                if (a instanceof Long && b instanceof Long) {
                    return integerFunc.applyAsLong((Long) a, (Long) b);
                }
            }
            Double x = LuaValue.toFloat(a);
            if (x != null) {
                Double y = LuaValue.toFloat(b);
                if (y != null) {
                    return floatFunc.applyAsDouble(x, y);
                }
            }
        }
        return null;
    }

}

File: code/python/ch07/src/arithmetic.py
from collections import namedtuple
from lua_value import LuaValue
from arith_op import ArithOp


def add(a, b):
    return a + b


def sub(a, b):
    return a - b


def mul(a, b):
    return a * b


def mod(a, b):
    return a % b


def div(a, b):
    return a / b


def fdiv(a, b):
    return a // b


def band(a, b):
    return a & b


def bor(a, b):
    return a | b


def bxor(a, b):
    return a ^ b


def shl(a, b):
    if b >= 0:
        return a << b
    return a >> (-b)


def shr(a, b):
    if b >= 0:
        return a >> b
    return a << (-b)


def unm(a, b):
    return -a + (b-b)


def bnot(a, b):
    return ~a + (b-b)


class Arithmetic:
    operator = namedtuple('operator', ['integer_func', 'float_func'])
    operators = {
        ArithOp.ADD:  operator(add,  add),
        ArithOp.SUB:  operator(sub,  sub),
        ArithOp.MUL:  operator(mul,  mul),
        ArithOp.MOD:  operator(mod,  mod),
        ArithOp.POW:  operator(None, pow),
        ArithOp.DIV:  operator(None, div),
        ArithOp.IDIV: operator(fdiv, fdiv),
        ArithOp.BAND: operator(band, None),
        ArithOp.BOR:  operator(bor,  None),
        ArithOp.BXOR: operator(bxor, None),
        ArithOp.SHL:  operator(shl,  None),
        ArithOp.SHR:  operator(shr,  None),
        ArithOp.UNM:  operator(unm,  unm),
        ArithOp.BNOT: operator(bnot, None)
    }

    @staticmethod
    def arith(a, op, b):
        integer_func = Arithmetic.operators[op].integer_func
        float_func = Arithmetic.operators[op].float_func

        if float_func is None:
            x = LuaValue.to_integer(a)
            if x:
                y = LuaValue.to_integer(b)
                if y:
                    return integer_func(x, y)
        else:
            if integer_func is not None:
                if isinstance(a, int) and isinstance(b, int):
                    return integer_func(int(a), int(b))
            x = LuaValue.to_float(a)
            if x:
                y = LuaValue.to_float(b)
                if y:
                    return float_func(x, y)

        return None
